### DISCUSSION TITLE
Pass function inside .then() consumer of Promise .call() in livewire using Global Livewire JavaScript Object

### DISCUSSION THREAD

**Discussion by aspyropoulos** [livewire/livewire#3286] [1. Help]
Hello,
I want to pass function **parseValue**() inside .then() consumer of Promise .call() in livewire ie

```
function parseValue(value) {
   console.log(value)
}

livewire.components.getComponentsByName('product')[0]
                .call('getVariants').then(function(value) {
                    parseValue(value)
                })
```

If that is not possible how else can I do this using the Global Livewire JavaScript Object?

**Comment by joshhanley** [comment]
@aspyropoulos you can't access the result of the promise as it's an internal function.

Can you describe what you are hoping to achieve and we can see if there is different way to do it?

Thanks!

**Comment by aspyropoulos** [comment]
hi @joshhanley,

I wanted to call a component's method by using livewire's javascript global object and get some data back. What I ended up doing is the use the global javascript object for calling the component's method then dispatch a browser event from within the component and finally add an event listener in my javascript for catching the event:

1.
`livewire.components.getComponentsByName('product')[0].call('getVariants')`

2.
```
class Product extends Component
{
  public function getVariants() {
    ...
    $this->dispatchBrowserEvent('variantsLoaded', ['data' => $data]);
  }
}
```

3.
```
function parseValue(value) {
   console.log(value)
}

window.addEventListener('variantsLoaded', event => {
   parseValue(event.detail.data)
})

```

Is there any better way than the above?

**Reply by joshhanley** [reply]
@aspyropoulos hmm it depends, what's the use-case for calling a method and getting data back? There are a few options, but it's context specific. If you can explain some of the reasons with a concrete example and the why it's needed we can work out if thats the best way 🙂

**Reply by aspyropoulos** [reply]
@joshhanley basically I am using [tiptap editor with LW](https://www.tiptap.dev/installation/livewire) and I am writing a custom plugin there  based on this [plugin](https://github.com/ueberdosis/tiptap/blob/main/packages/extension-bubble-menu/src/bubble-menu-plugin.ts), that sets/updates custom highlighted marks inside the text editor. To do that once my editor is set up I am doing a livewire method call to fetch the data and then parse them to update the highlighted values. 

A simple example to show the case is:

In the constructor of the above plugin I call my components `getVariants `method (step 1 from my previous answer) plus
an event listener to fetch event response from my component (step 3 from my previous answer). Let me know if this is enough to show the case.
